lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -244,6 +244,90 @@ TypeChecker::lookupUnqualifiedType(DeclContext *dc, DeclName name,
 
+  // If we're looking for a type named `CodingKeys`, then we've got some
+  // additional work to do here. CodingKeys is a special type: although not a
+  // protocol requirement of the Encodable/Decodable protocols, it is
+  // synthesized as part of Encodable/Decodable conformance. Unqualified lookup
+  // may occur before the type is synthesized, though, in which case, we need to
+  // make sure it gets synthesized.
+  //
+  // The unqualified lookups above, however, cannot force Encodable/Decodable
+  // synthesis to occur (since they know nothing about typechecking). Thus, we
+  // also need to find the closest qualified type which might contain a
+  // CodingKeys type and synthesize it.
+  //
+  // If this type is "closer" than the closest unqualified type, then we can
+  // use it.
+  if (name.isSimpleName(Context.Id_CodingKeys)) {
+    SmallVector<TypeDecl *, 1> qualifiedDecls;
+
+    auto *currentDC = dc;
+    while (currentDC) {
+      auto *nominal =
+        currentDC->getAsNominalTypeOrNominalTypeExtensionContext();
+      if (nominal) {
+        // We've found a nominal context to look in. Let's preform a qualified
+        // lookup here (which will cause CodingKeys synthesis if possible).
+        auto lookupOptions = options | NameLookupFlags::ProtocolMembers;
+        auto results = lookupMemberType(nominal,
+                                        nominal->getDeclaredInterfaceType(),
+                                        Context.Id_CodingKeys, lookupOptions);
+        if (results.size() > 0) {
+          for (auto &result : results) {
+            qualifiedDecls.push_back(result.first);
+          }
+
+          break;
+        }
+      }
+
+      currentDC = currentDC->getParent();
+    }
+
+    // If there were any results, we need to merge them with any decls already
+    // found by the unqualified lookup above. Those decls might take precedence
+    // based on their locality.
+    auto insertionIndex = decls.size();
+    decls.insert(decls.begin(), qualifiedDecls.begin(), qualifiedDecls.end());
+
+    bool hasDups = false;
+    if (insertionIndex > 0 && decls.size() != insertionIndex) {
+      // Both arrays contained elements. We need to merge the results.
+      std::inplace_merge(decls.begin(), decls.begin() + insertionIndex,
+                         decls.end(),
+                         [&hasDups](const TypeDecl *d1, const TypeDecl *d2) {
+        if (d1 == d2) {
+          hasDups = true;
+          return true;
+        }
+
+        auto *dc1 = dyn_cast<DeclContext>(d1);
+        auto *dc2 = dyn_cast<DeclContext>(d2);
+        if (!dc1 || !dc2) {
+          // We can't really compare these if either does not evaluate to a
+          // DeclContext. Keep this sort stable at least.
+          return true;
+        }
+
+        // If d1 and d2 share a parent DeclContext, then they are equally
+        // "close". Otherwise, if d1 is a child of d2's parent context, then
+        // it must be "closer" than d2.
```

Comment:
That's the unfortunate thing — we can't really tell for sure whether the results of normal unqualified lookup are absolutely correct, since we don't know if the contexts the lookup has iterated have gone through synthesis yet or not. Since the lookup is basically a black box, it looks like we unfortunately have to go through and do this anyway at some point. I think a refactor of unqualified lookups to insert hooks is probably a bigger change than what we want at this point.

Let me look into extensions and see what we can do.